modules/tinymce/src/core/main/ts/keyboard/NavigationUtils.ts

```diff
@@ -72,13 +78,21 @@ const moveVertically = (editor: Editor, direction: LineWalker.VDirection, range:
   const caretPosition = CaretUtils.getNormalizedRangeEndPoint(direction, editor.getBody(), range);
   const caretClientRect = ArrUtils.last(caretPosition.getClientRects());
   const forwards = direction === LineWalker.VDirection.Down;
+  const root = editor.getBody();
 
   if (!caretClientRect) {
     return Optional.none();
   }
 
+  if (isCefAtEdgeSelected(editor)) {
+    const caretPosition = forwards ? CaretPosition.fromRangeEnd(range) : CaretPosition.fromRangeStart(range);
+    return (!forwards ? LineReader.getClosestPositionAbove(root, caretPosition) : LineReader.getClosestPositionBelow(root, caretPosition))
+      .or(Optional.from(caretPosition))
```

Comment:
This is going to eagerly evaluate and create the Optional when not needed, so this should be using `orThunk`.
```suggestion
      .orThunk(() => Optional.from(caretPosition))
```